community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/ShortestPath.java

```diff
@@ -197,14 +193,21 @@ public boolean equals( Object obj )
         }
     }
 
-    private void goOneStep( DirectionData directionData, DirectionData otherSide, Hits hits,
-            DirectionData startSide, boolean stopAsap )
+    private void goOneStep( DirectionData directionData, DirectionData otherSide, Hits hits, DirectionData startSide,
+            boolean stopAsap )
     {
         if ( !directionData.hasNext() )
         {
+            // We can not go any deeper from this direction. Possibly disconnected nodes.
+            directionData.isExhausted = true;
             return;
         }
-        
+        if ( otherSide.isExhausted )
+        {
+            // Other side can not go any deeper.
+            // If this side do not find a path in this layer, the nodes are disconnected and we can stop the search.
+            directionData.finishCurrentLayerThenStop = true;
```

Comment:
To me it looks like `finishCurrentLayerThenStop` is a direct function of the other side `isExhausted=true`, can't we use that instead of having two variables for the same thing? At least we could set `otherSise.finishCurrentLayerThenStop = true` right under `directionData.isExhausted = true` above, to avoid checking this boolean for every single step.